# --- T2-COPYRIGHT-BEGIN ---
# t2/package/*/linux/revert-sunhme.patch
# Copyright (C) 2026 The T2 SDE Project
# SPDX-License-Identifier: GPL-2.0 or patched project license
# --- T2-COPYRIGHT-END ---

hme f0072f38: error -EBUSY: can't request region for resource [mem 0x1ffe8c07000-0x1ffe8c0701f]
hme f0072f38: Cannot map TCVR registers.
hme f0072f38: probe with driver hme failed with error -16
hme f007ab44: error -EBUSY: can't request region for resource [mem 0x1ff28c07000-0x1ff28c0701f]
hme f007ab44: Cannot map TCVR registers.
hme f007ab44: probe with driver hme failed with error -16

1ff28c00000-1ff28c00107 : HME Global Regs
1ff28c02000-1ff28c02033 : HME TX Regs
1ff28c04000-1ff28c0401f : HME RX Regs
1ff28c06000-1ff28c0635f : HME BIGMAC Regs
1ff28c07000-1ff28c0701f : HME Tranceiver Regs

00 00 00 02 08 c0 00 00  00 00 01 08
00 00 00 02 08 c0 20 00  00 00 20 00
00 00 00 02 08 c0 40 00  00 00 20 00
00 00 00 02 08 c0 60 00  00 00 20 00
00 00 00 02 08 c0 70 00  00 00 00 20

--- linux-6.18/drivers/net/ethernet/sun/sunhme.c.orig	2025-11-30 23:42:10.000000000 +0100
+++ linux-6.18/drivers/net/ethernet/sun/sunhme.c	2026-02-04 21:28:08.869823984 +0100
@@ -2524,45 +2524,46 @@
 
 	spin_lock_init(&hp->happy_lock);
 
+	err = -ENODEV;
 	if (qp != NULL) {
 		hp->qfe_parent = qp;
 		hp->qfe_ent = qfe_slot;
 		qp->happy_meals[qfe_slot] = dev;
 	}
 
-	hp->gregs = devm_platform_ioremap_resource(op, 0);
-	if (IS_ERR(hp->gregs)) {
+	hp->gregs = of_ioremap(&op->resource[0], 0,
+			       GREG_REG_SIZE, "HME Global Regs");
+	if (!hp->gregs) {
 		dev_err(&op->dev, "Cannot map global registers.\n");
-		err = PTR_ERR(hp->gregs);
-		goto err_out_clear_quattro;
+		goto err_out;
 	}
 
-	hp->etxregs = devm_platform_ioremap_resource(op, 1);
-	if (IS_ERR(hp->etxregs)) {
+	hp->etxregs = of_ioremap(&op->resource[1], 0,
+				 ETX_REG_SIZE, "HME TX Regs");
+	if (!hp->etxregs) {
 		dev_err(&op->dev, "Cannot map MAC TX registers.\n");
-		err = PTR_ERR(hp->etxregs);
-		goto err_out_clear_quattro;
+		goto err_out_iounmap;
 	}
 
-	hp->erxregs = devm_platform_ioremap_resource(op, 2);
-	if (IS_ERR(hp->erxregs)) {
+	hp->erxregs = of_ioremap(&op->resource[2], 0,
+				 ERX_REG_SIZE, "HME RX Regs");
+	if (!hp->erxregs) {
 		dev_err(&op->dev, "Cannot map MAC RX registers.\n");
-		err = PTR_ERR(hp->erxregs);
-		goto err_out_clear_quattro;
+		goto err_out_iounmap;
 	}
 
-	hp->bigmacregs = devm_platform_ioremap_resource(op, 3);
-	if (IS_ERR(hp->bigmacregs)) {
+	hp->bigmacregs = of_ioremap(&op->resource[3], 0,
+				    BMAC_REG_SIZE, "HME BIGMAC Regs");
+	if (!hp->bigmacregs) {
 		dev_err(&op->dev, "Cannot map BIGMAC registers.\n");
-		err = PTR_ERR(hp->bigmacregs);
-		goto err_out_clear_quattro;
+		goto err_out_iounmap;
 	}
 
-	hp->tcvregs = devm_platform_ioremap_resource(op, 4);
-	if (IS_ERR(hp->tcvregs)) {
+	hp->tcvregs = of_ioremap(&op->resource[4], 0,
+				 TCVR_REG_SIZE, "HME Tranceiver Regs");
+	if (!hp->tcvregs) {
 		dev_err(&op->dev, "Cannot map TCVR registers.\n");
-		err = PTR_ERR(hp->tcvregs);
-		goto err_out_clear_quattro;
+		goto err_out_iounmap;
 	}
 
 	hp->hm_revision = 0xa0;
@@ -2604,6 +2605,21 @@
 err_out_clear_quattro:
 	if (qp)
 		qp->happy_meals[qfe_slot] = NULL;
+
+err_out_iounmap:
+	if (hp->gregs)
+		of_iounmap(&op->resource[0], hp->gregs, GREG_REG_SIZE);
+	if (hp->etxregs)
+		of_iounmap(&op->resource[1], hp->etxregs, ETX_REG_SIZE);
+	if (hp->erxregs)
+		of_iounmap(&op->resource[2], hp->erxregs, ERX_REG_SIZE);
+	if (hp->bigmacregs)
+		of_iounmap(&op->resource[3], hp->bigmacregs, BMAC_REG_SIZE);
+	if (hp->tcvregs)
+		of_iounmap(&op->resource[4], hp->tcvregs, TCVR_REG_SIZE);
+
+err_out:
+
 	return err;
 }
 #endif
@@ -2818,6 +2834,22 @@
 	return happy_meal_sbus_probe_one(op, is_qfe);
 }
 
+static void hme_sbus_remove(struct platform_device *op)
+{
+	struct happy_meal *hp = platform_get_drvdata(op);
+	//struct net_device *net_dev = hp->dev;
+
+	//unregister_netdev(net_dev);
+
+	of_iounmap(&op->resource[0], hp->gregs, GREG_REG_SIZE);
+	of_iounmap(&op->resource[1], hp->etxregs, ETX_REG_SIZE);
+	of_iounmap(&op->resource[2], hp->erxregs, ERX_REG_SIZE);
+	of_iounmap(&op->resource[3], hp->bigmacregs, BMAC_REG_SIZE);
+	of_iounmap(&op->resource[4], hp->tcvregs, TCVR_REG_SIZE);
+
+	//free_netdev(net_dev);
+}
+
 static const struct of_device_id hme_sbus_match[] = {
 	{
 		.name = "SUNW,hme",
@@ -2841,6 +2873,7 @@
 		.of_match_table = hme_sbus_match,
 	},
 	.probe		= hme_sbus_probe,
+	.remove		= hme_sbus_remove,
 };
 
 static int __init happy_meal_sbus_init(void)
